routers/user/auth.go

```diff
@@ -649,27 +644,76 @@ func SignInOAuthCallback(ctx *context.Context) {
 				LoginName:   gothUser.UserID,
 			}
 
-			if !createAndHandleCreatedUser(ctx, base.TplName(""), nil, u, &gothUser) {
+			if ok, needLink := createAndHandleCreatedUser(ctx, base.TplName(""), nil, u, &gothUser, setting.OAuth2Client.AccountLinkingLogin || setting.OAuth2Client.AutoRegistrationLinking); !ok {
 				// error already handled
+				if needLink {
+					if setting.OAuth2Client.AutoRegistrationLinking {
+						var user *models.User
+						user = &models.User{Name: u.Name}
+						hasUser, err := models.GetUser(user)
+						if !hasUser || err != nil {
+							user = &models.User{Email: u.Email}
+							hasUser, err = models.GetUser(user)
+							if !hasUser || err != nil {
+								ctx.ServerError("UserLinkAccount", err)
+								return
+							}
+						}
+
+						// TODO: probably we should respect user's choice...
+						linkAccount(ctx, user, gothUser, true)
+					} else {
+						showLinkingLogin(ctx, gothUser)
+					}
+				}
 				return
 			}
 		} else {
 			// no existing user is found, request attach or new account
-			if err := ctx.Session.Set("linkAccountGothUser", gothUser); err != nil {
-				log.Error("Error setting linkAccountGothUser in session: %v", err)
-			}
-			if err := ctx.Session.Release(); err != nil {
-				log.Error("Error storing session: %v", err)
-			}
-			ctx.Redirect(setting.AppSubURL + "/user/link_account")
-			return
+			showLinkingLogin(ctx, gothUser)
 		}
 	}
 
 	handleOAuth2SignIn(ctx, u, gothUser)
 }
 
+func getUserName(gothUser *goth.User) string {
+	if setting.OAuth2Client.UseEmail {
+		return strings.Split(gothUser.Email, "@")[0]
+	} else if setting.OAuth2Client.UseNickname {
+		return gothUser.NickName
+	} else {
+		return gothUser.UserID
+	}
+}
+
+func showLinkingLogin(ctx *context.Context, gothUser goth.User) {
+	if err := ctx.Session.Set("linkAccountGothUser", gothUser); err != nil {
+		log.Error("Error setting linkAccountGothUser in session: %v", err)
+	}
+	if err := ctx.Session.Release(); err != nil {
+		log.Error("Error storing session: %v", err)
+	}
+	ctx.Redirect(setting.AppSubURL + "/user/link_account")
+	return
+}
+
+func updateAvatarIfNeed(url string, u *models.User) {
+	if setting.OAuth2Client.UpdateAvatar && len(url) > 0 {
+		resp, err := http.Get(url)
+		// ignore any error
+		if err == nil && resp.StatusCode == http.StatusOK {
+			data, err := io.ReadAll(resp.Body)
+			if err == nil {
```

Comment:
We should respect the `setting.Avatar.MaxFileSize`:
```suggestion
			data, err := io.ReadAll(io.LimitReader(resp.Body, setting.Avatar.MaxFileSize + 1))
			if err == nil && len(data) <= setting.Avatar.MaxFileSize {
```